### ISSUE TITLE
ValueError specifying TensorSpec names while using model.save

### ISSUE THREAD

**Issue by rlav440** [tensorflow/tensorflow#38465]
**System information** 
- Custom code has been written to define the model, and is given
- OS Platform and Distribution: Linux Ubuntu 19.10: 
- TensorFlow installed via pip
- TensorFlow versions 2.1.0 & 2.2.0.dev20200411 trialed
- Python version: 3.7

**Describe the current behavior**
Attempting to use model.save() to save a trained model, raises the eventual error 
`ValueError: If specifying TensorSpec names for nested structures, either zero or all names have to be specified.`
Code also raises the error while failing to save
`WARNING:tensorflow:AutoGraph could not transform <function canonicalize_signatures.<locals>.signature_wrapper at 0x7efb5076e0e0> and will run it as-is.
Please report this to the TensorFlow team. When filing the bug, set the verbosity to 10 (on Linux, `export AUTOGRAPH_VERBOSITY=10`) and attach the full output.
Cause: Bad argument number for Name: 4, expecting 3`
Another error that visible in the output is the failed conversion to JSON of a tf_graphics layer. This has been raised in the tensorflow_graphics repository

**Describe the expected behaviour**
Unsure - error was asked to be raised to TensorFlow team by the error text.

**Standalone code to reproduce the issue** 
In comment below

**Current Code output**
```
WARNING:tensorflow:AutoGraph could not transform <bound method FeatureSteeredConvolutionKerasLayer.call of <tensorflow_graphics.nn.layer.graph_convolution.FeatureSteeredConvolutionKerasLayer object at 0x7efb9617c790>> and will run it as-is.
Please report this to the TensorFlow team. When filing the bug, set the verbosity to 10 (on Linux, `export AUTOGRAPH_VERBOSITY=10`) and attach the full output.
Cause: module 'gast' has no attribute 'Constant'
To silence this warning, decorate the function with @tf.autograph.experimental.do_not_convert
Model: "MyFirstGraphModel"
__________________________________________________________________________________________________
Layer (type)                    Output Shape         Param #     Connected to                     
==================================================================================================
input_1 (InputLayer)            [(None, 100, 3)]     0                                            
__________________________________________________________________________________________________
conv1d (Conv1D)                 (None, 100, 2)       8           input_1[0][0]                    
__________________________________________________________________________________________________
input_2 (InputLayer)            [(None, 100, 100)]   0                                            
__________________________________________________________________________________________________
input_3 (InputLayer)            [(None,)]            0                                            
__________________________________________________________________________________________________
feature_steered_convolution_ker (None, 100, 32)      584         conv1d[0][0]                     
                                                                 input_2[0][0]                    
__________________________________________________________________________________________________
re_lu (ReLU)                    (None, 100, 32)      0           feature_steered_convolution_keras
__________________________________________________________________________________________________
feature_steered_convolution_ker (None, 100, 64)      16968       re_lu[0][0]                      
                                                                 input_2[0][0]                    
__________________________________________________________________________________________________
re_lu_1 (ReLU)                  (None, 100, 64)      0           feature_steered_convolution_keras
__________________________________________________________________________________________________
feature_steered_convolution_ker (None, 100, 128)     66696       re_lu_1[0][0]                    
                                                                 input_2[0][0]                    
__________________________________________________________________________________________________
re_lu_2 (ReLU)                  (None, 100, 128)     0           feature_steered_convolution_keras
__________________________________________________________________________________________________
conv1d_1 (Conv1D)               (None, 100, 256)     33024       re_lu_2[0][0]                    
__________________________________________________________________________________________________
conv1d_2 (Conv1D)               (None, 100, 1)       257         conv1d_1[0][0]                   
==================================================================================================
Total params: 117,537
Trainable params: 117,537
Non-trainable params: 0
__________________________________________________________________________________________________
WARNING:tensorflow:AutoGraph could not transform <function tf_record_to_dataset.<locals>.<lambda> at 0x7efb9450be60> and will run it as-is.
Please report this to the TensorFlow team. When filing the bug, set the verbosity to 10 (on Linux, `export AUTOGRAPH_VERBOSITY=10`) and attach the full output.
Cause: module 'gast' has no attribute 'Constant'
To silence this warning, decorate the function with @tf.autograph.experimental.do_not_convert
WARNING:tensorflow:AutoGraph could not transform <function tf_record_to_dataset.<locals>.<lambda> at 0x7efb9448ab00> and will run it as-is.
Please report this to the TensorFlow team. When filing the bug, set the verbosity to 10 (on Linux, `export AUTOGRAPH_VERBOSITY=10`) and attach the full output.
Cause: module 'gast' has no attribute 'Constant'
To silence this warning, decorate the function with @tf.autograph.experimental.do_not_convert
2020-04-12 15:51:25.362023: I tensorflow/core/profiler/lib/profiler_session.cc:154] Profiler session started.
2020-04-12 15:51:25.362060: I tensorflow/core/profiler/internal/gpu/cupti_tracer.cc:1372] Profiler found 1 GPUs
2020-04-12 15:51:25.362157: W tensorflow/stream_executor/platform/default/dso_loader.cc:55] Could not load dynamic library 'libcupti.so.10.1'; dlerror: libcupti.so.10.1: cannot open shared object file: No such file or directory
2020-04-12 15:51:25.362163: E tensorflow/core/profiler/internal/gpu/cupti_tracer.cc:1419] function cupti_interface_->Subscribe( &subscriber_, (CUpti_CallbackFunc)ApiCallback, this)failed with error CUPTI could not be loaded or symbol could not be found.
2020-04-12 15:51:25.362167: E tensorflow/core/profiler/internal/gpu/cupti_tracer.cc:1458] function cupti_interface_->ActivityRegisterCallbacks( AllocCuptiActivityBuffer, FreeCuptiActivityBuffer)failed with error CUPTI could not be loaded or symbol could not be found.
2020-04-12 15:51:25.362175: E tensorflow/core/profiler/internal/gpu/cupti_tracer.cc:1441] function cupti_interface_->EnableCallback( 0 , subscriber_, CUPTI_CB_DOMAIN_DRIVER_API, cbid)failed with error CUPTI could not be loaded or symbol could not be found.
2020-04-12 15:51:25.362203: I tensorflow/core/profiler/lib/profiler_session.cc:154] Profiler session started.
2020-04-12 15:51:25.362208: E tensorflow/core/profiler/internal/gpu/cupti_tracer.cc:1419] function cupti_interface_->Subscribe( &subscriber_, (CUpti_CallbackFunc)ApiCallback, this)failed with error CUPTI could not be loaded or symbol could not be found.
2020-04-12 15:51:25.362211: E tensorflow/core/profiler/internal/gpu/cupti_tracer.cc:1458] function cupti_interface_->ActivityRegisterCallbacks( AllocCuptiActivityBuffer, FreeCuptiActivityBuffer)failed with error CUPTI could not be loaded or symbol could not be found.
2020-04-12 15:51:25.362216: E tensorflow/core/profiler/internal/gpu/cupti_tracer.cc:1441] function cupti_interface_->EnableCallback( 0 , subscriber_, CUPTI_CB_DOMAIN_DRIVER_API, cbid)failed with error CUPTI could not be loaded or symbol could not be found.
WARNING:tensorflow:Model failed to serialize as JSON. Ignoring... Layer FeatureSteeredConvolutionKerasLayer has arguments in `__init__` and therefore must override `get_config`.
WARNING:tensorflow:AutoGraph could not transform <function Model.make_train_function.<locals>.train_function at 0x7efb944d2560> and will run it as-is.
Please report this to the TensorFlow team. When filing the bug, set the verbosity to 10 (on Linux, `export AUTOGRAPH_VERBOSITY=10`) and attach the full output.
Cause: Bad argument number for Name: 4, expecting 3
To silence this warning, decorate the function with @tf.autograph.experimental.do_not_convert
/home/robin/anaconda3/envs/Robin-Graph-NN-Pose-from-Model/lib/python3.7/site-packages/tensorflow/python/framework/indexed_slices.py:430: UserWarning: Converting sparse IndexedSlices to a dense Tensor of unknown shape. This may consume a large amount of memory.
  "Converting sparse IndexedSlices to a dense Tensor of unknown shape. "
2020-04-12 15:51:27.041239: I tensorflow/core/profiler/lib/profiler_session.cc:154] Profiler session started.
2020-04-12 15:51:27.041279: E tensorflow/core/profiler/internal/gpu/cupti_tracer.cc:1419] function cupti_interface_->Subscribe( &subscriber_, (CUpti_CallbackFunc)ApiCallback, this)failed with error CUPTI could not be loaded or symbol could not be found.
2020-04-12 15:51:27.041288: E tensorflow/core/profiler/internal/gpu/cupti_tracer.cc:1458] function cupti_interface_->ActivityRegisterCallbacks( AllocCuptiActivityBuffer, FreeCuptiActivityBuffer)failed with error CUPTI could not be loaded or symbol could not be found.
WARNING:tensorflow:From /home/robin/anaconda3/envs/Robin-Graph-NN-Pose-from-Model/lib/python3.7/site-packages/tensorflow/python/ops/summary_ops_v2.py:1271: stop (from tensorflow.python.eager.profiler) is deprecated and will be removed after 2020-07-01.
Instructions for updating:
use `tf.profiler.experimental.stop` instead.
2020-04-12 15:51:27.046579: E tensorflow/core/profiler/internal/gpu/cupti_tracer.cc:1441] function cupti_interface_->EnableCallback( 0 , subscriber_, CUPTI_CB_DOMAIN_DRIVER_API, cbid)failed with error CUPTI could not be loaded or symbol could not be found.
2020-04-12 15:51:27.049348: I tensorflow/core/profiler/internal/gpu/device_tracer.cc:217]  GpuTracer has collected 0 callback api events and 0 activity events.
2020-04-12 15:51:27.054982: I tensorflow/core/profiler/rpc/client/save_profile.cc:168] Creating directory: logs/fit/20200412-155125/train/plugins/profile/2020_04_12_15_51_27
      1/Unknown - 0s 81us/step - loss: 0.3479 - acc: 0.69002020-04-12 15:51:27.059084: I tensorflow/core/profiler/rpc/client/save_profile.cc:174] Dumped gzipped tool data for trace.json.gz to logs/fit/20200412-155125/train/plugins/profile/2020_04_12_15_51_27/robin-System-Product-Name.trace.json.gz
2020-04-12 15:51:27.061354: I tensorflow/core/profiler/utils/event_span.cc:288] Generation of step-events took 0 ms

2020-04-12 15:51:27.063255: I tensorflow/python/profiler/internal/profiler_wrapper.cc:91] Creating directory: logs/fit/20200412-155125/train/plugins/profile/2020_04_12_15_51_27Dumped tool data for overview_page.pb to logs/fit/20200412-155125/train/plugins/profile/2020_04_12_15_51_27/robin-System-Product-Name.overview_page.pb
Dumped tool data for input_pipeline.pb to logs/fit/20200412-155125/train/plugins/profile/2020_04_12_15_51_27/robin-System-Product-Name.input_pipeline.pb
Dumped tool data for tensorflow_stats.pb to logs/fit/20200412-155125/train/plugins/profile/2020_04_12_15_51_27/robin-System-Product-Name.tensorflow_stats.pb
Dumped tool data for kernel_stats.pb to logs/fit/20200412-155125/train/plugins/profile/2020_04_12_15_51_27/robin-System-Product-Name.kernel_stats.pb

    628/Unknown - 2s 3ms/step - loss: 0.3358 - acc: 0.6680WARNING:tensorflow:AutoGraph could not transform <function Model.make_test_function.<locals>.test_function at 0x7efb602a50e0> and will run it as-is.
Please report this to the TensorFlow team. When filing the bug, set the verbosity to 10 (on Linux, `export AUTOGRAPH_VERBOSITY=10`) and attach the full output.
Cause: Bad argument number for Name: 4, expecting 3
To silence this warning, decorate the function with @tf.autograph.experimental.do_not_convert
640/640 [==============================] - 2s 4ms/step - loss: 0.3356 - acc: 0.6681 - val_loss: 0.3316 - val_acc: 0.6689
2020-04-12 15:51:30.035679: W tensorflow/python/util/util.cc:329] Sets are not currently considered sequences, but this may change in the future, so consider avoiding using them.
WARNING:tensorflow:AutoGraph could not transform <function canonicalize_signatures.<locals>.signature_wrapper at 0x7efb5076e0e0> and will run it as-is.
Please report this to the TensorFlow team. When filing the bug, set the verbosity to 10 (on Linux, `export AUTOGRAPH_VERBOSITY=10`) and attach the full output.
Cause: Bad argument number for Name: 4, expecting 3
To silence this warning, decorate the function with @tf.autograph.experimental.do_not_convert
Traceback (most recent call last):
  File "/home/robin/Masters/Robin-Graph-NN-Pose-from-Model/Examples/graph_network_model.py", line 68, in <module>
    model.save('./test/model_dir')
  File "/home/robin/anaconda3/envs/Robin-Graph-NN-Pose-from-Model/lib/python3.7/site-packages/tensorflow/python/keras/engine/network.py", line 1062, in save
    signatures, options)
  File "/home/robin/anaconda3/envs/Robin-Graph-NN-Pose-from-Model/lib/python3.7/site-packages/tensorflow/python/keras/saving/save.py", line 134, in save_model
    signatures, options)
  File "/home/robin/anaconda3/envs/Robin-Graph-NN-Pose-from-Model/lib/python3.7/site-packages/tensorflow/python/keras/saving/saved_model/save.py", line 78, in save
    save_lib.save(model, filepath, signatures, options)
  File "/home/robin/anaconda3/envs/Robin-Graph-NN-Pose-from-Model/lib/python3.7/site-packages/tensorflow/python/saved_model/save.py", line 951, in save
    obj, export_dir, signatures, options, meta_graph_def)
  File "/home/robin/anaconda3/envs/Robin-Graph-NN-Pose-from-Model/lib/python3.7/site-packages/tensorflow/python/saved_model/save.py", line 1024, in _build_meta_graph
    _ = _SaveableView(checkpoint_graph_view)
  File "/home/robin/anaconda3/envs/Robin-Graph-NN-Pose-from-Model/lib/python3.7/site-packages/tensorflow/python/saved_model/save.py", line 204, in __init__
    function._list_all_concrete_functions_for_serialization())  # pylint: disable=protected-access
  File "/home/robin/anaconda3/envs/Robin-Graph-NN-Pose-from-Model/lib/python3.7/site-packages/tensorflow/python/eager/def_function.py", line 882, in _list_all_concrete_functions_for_serialization
    concrete_functions.append(self.get_concrete_function(*args, **kwargs))
  File "/home/robin/anaconda3/envs/Robin-Graph-NN-Pose-from-Model/lib/python3.7/site-packages/tensorflow/python/keras/saving/saved_model/save_impl.py", line 546, in get_concrete_function
    self.call_collection.add_trace(*args, **kwargs)
  File "/home/robin/anaconda3/envs/Robin-Graph-NN-Pose-from-Model/lib/python3.7/site-packages/tensorflow/python/keras/saving/saved_model/save_impl.py", line 418, in add_trace
    trace_with_training(True)
  File "/home/robin/anaconda3/envs/Robin-Graph-NN-Pose-from-Model/lib/python3.7/site-packages/tensorflow/python/keras/saving/saved_model/save_impl.py", line 416, in trace_with_training
    fn.get_concrete_function(*args, **kwargs)
  File "/home/robin/anaconda3/envs/Robin-Graph-NN-Pose-from-Model/lib/python3.7/site-packages/tensorflow/python/keras/saving/saved_model/save_impl.py", line 547, in get_concrete_function
    return super(LayerCall, self).get_concrete_function(*args, **kwargs)
  File "/home/robin/anaconda3/envs/Robin-Graph-NN-Pose-from-Model/lib/python3.7/site-packages/tensorflow/python/eager/def_function.py", line 1000, in get_concrete_function
    concrete = self._get_concrete_function_garbage_collected(*args, **kwargs)
  File "/home/robin/anaconda3/envs/Robin-Graph-NN-Pose-from-Model/lib/python3.7/site-packages/tensorflow/python/eager/def_function.py", line 918, in _get_concrete_function_garbage_collected
    *args, **kwargs)
  File "/home/robin/anaconda3/envs/Robin-Graph-NN-Pose-from-Model/lib/python3.7/site-packages/tensorflow/python/eager/function.py", line 2497, in _get_concrete_function_garbage_collected
    graph_function, args, kwargs = self._maybe_define_function(args, kwargs)
  File "/home/robin/anaconda3/envs/Robin-Graph-NN-Pose-from-Model/lib/python3.7/site-packages/tensorflow/python/eager/function.py", line 2775, in _maybe_define_function
    graph_function = self._create_graph_function(args, kwargs)
  File "/home/robin/anaconda3/envs/Robin-Graph-NN-Pose-from-Model/lib/python3.7/site-packages/tensorflow/python/eager/function.py", line 2665, in _create_graph_function
    capture_by_value=self._capture_by_value),
  File "/home/robin/anaconda3/envs/Robin-Graph-NN-Pose-from-Model/lib/python3.7/site-packages/tensorflow/python/framework/func_graph.py", line 898, in func_graph_from_py_func
    args, arg_names, flat_shapes=arg_shapes)
  File "/home/robin/anaconda3/envs/Robin-Graph-NN-Pose-from-Model/lib/python3.7/site-packages/tensorflow/python/framework/func_graph.py", line 1134, in _get_defun_inputs_from_args
    args, names, structure=args, flat_shapes=flat_shapes)
  File "/home/robin/anaconda3/envs/Robin-Graph-NN-Pose-from-Model/lib/python3.7/site-packages/tensorflow/python/framework/func_graph.py", line 1198, in _get_defun_inputs
    raise ValueError("If specifying TensorSpec names for nested structures, "
ValueError: If specifying TensorSpec names for nested structures, either zero or all names have to be specified.

```

**Comment by rlav440** [comment]
Attached is the dataset with which this error was raised.
Apologies for the initial long code!
[Minimal_code.zip](https://github.com/tensorflow/tensorflow/files/4465842/Minimal_code.zip)
[Dataset1.zip](https://github.com/tensorflow/tensorflow/files/4465837/Dataset1.zip)

**Comment by ravikyram** [comment]
@rlav440 

Looks like code is incomplete. I am seeing the error `ModuleNotFoundError: No module named 'tensorflow_graphics'`. Please, help us with colab link or simple standalone code to reproduce the issue in our environment. It helps us in localizing the issue faster. Thanks!

**Comment by rlav440** [comment]
@ravikyram 
That's the tensorflow graphics api from:
https://github.com/tensorflow/graphics - a newer API for tensorflow.
Please note I build that by pulling from source (git link above) as there is currently a bug in the released pypi repository ver.

As I mentioned above, the model does save without the tensorflow graphics API components(I've raised an issue on that repository as well) but it still raises the same `Warnings:AutoGraph could not transform` so I have left this issue open with the reason it specifically requests the issue be reported.

**Comment by gowthamkpr** [comment]
@rlav440 This is similar to the issue [here](https://github.com/tensorflow/graphics/issues/227).  As mentioned in the issue, Removing the graph convolution layers allows the model to save.

Please take a look at it.

**Comment by rlav440** [comment]
Hi @gowthamkpr That is the issue that I mentioned posting. Upon looking at the first post, I believe I could have been clearer so have edited the first post to reflect the current nature of the issue

**Comment by woodshop** [comment]
I have run into the same error on a model built using the keras functional API. In particular, the error occurs when saving a checkpoint by calling `tfk.models.save_model`. My model takes two inputs (using the `tf.data` API), one of which is a `tf.RaggedTensor` and the other of which is a `tf.Tensor`. 

A `tf.RaggedTensorSpec` doesn't have a `name` attribute and at some point the `tf.RaggedTensorSpec` is converted by TF into several `tf.TensorSpec`s, which therefore also don't have values for their `name` attributes.

The error does not occur when saving a weights-only checkpoint with `tfk.Model.save_weights`.

**Comment by huonw** [comment]
I filed #40373 separately and was referred to this as a duplicate. It seems like there's not a standalone reproducer listed here yet. Here's the one using `SparseTensor` from that issue:

```python
import tensorflow as tf

class SqueezedSparseConversion(tf.keras.layers.Layer):
    def call(self, inputs):
        return tf.SparseTensor([(0, 1)], [0.1], (3, 3))

class GraphConvolution(tf.keras.layers.Layer):
    def call(self, inputs):
        return inputs[0]

x_t = tf.keras.Input(0)
sp = SqueezedSparseConversion()(x_t)
out = GraphConvolution()([x_t, sp])

m = tf.keras.Model([x_t], out)
m.summary()
m.save("")
```

Colab link: https://colab.research.google.com/gist/Saduf2019/56d0bb4e22fe010dbd150c6a8def9a54/untitled.ipynb

As I mentioned on the issue above, a helpful improvement here (potentially easier than fixing the whole issue) would be to add more info to the error message, such as which layer(s) are causing the problem, to make it easier to narrow down the issue.

**Comment by gowthamkpr** [comment]
@huonw @rlav440 This issue has been fixed in tf-nightly. Please find the gist [here](https://colab.research.google.com/gist/gowthamkpr/02054044d39cd7986950ee10fa740b10/untitled.ipynb)

**Comment by huonw** [comment]
I tested this on our real code in https://github.com/stellargraph/stellargraph/pull/1689 and the tests in question passed, so I can confirm that this seems to be fixed for our particular example. Thanks!

**Comment by gowthamkpr** [comment]
Closing this issue as it has been fixed in Nightly. Thanks!

**Comment by ashahab** [comment]
@gowthamkpr Do you have links to the PRs that fixed the issue? We are thinking of backporting the fixes to 2.2, and would love your help in finding the changes.

**Comment by ashahab** [comment]
@gowthamkpr can you help us locate the PR that fixes the issue?
